iree/tools/iree-run-trace-main.c

```diff
@@ -0,0 +1,750 @@
+                                    iree_make_cstring_view("module_load"))) {
+    return iree_trace_replay_event_module_load(replay, document, event_node);
+  } else if (iree_yaml_string_equal(type_node,
+                                    iree_make_cstring_view("call"))) {
+    return iree_trace_replay_event_call(replay, document, event_node);
+  }
+  return iree_make_status(
+      IREE_STATUS_UNIMPLEMENTED, "(%zu): unhandled type '%.*s'",
+      event_node->start_mark.line, (int)type_node->data.scalar.length,
+      type_node->data.scalar.value);
+}
+
+// Runs the trace in |file| using |root_path| as the base for any path lookups
+// required for external files referenced in |file|.
+static iree_status_t iree_run_trace_file(iree_string_view_t root_path,
+                                         FILE* file,
+                                         iree_vm_instance_t* instance) {
+  iree_trace_replay_t replay;
+  IREE_RETURN_IF_ERROR(iree_trace_replay_initialize(
+      root_path, instance, iree_allocator_system(), &replay));
+
+  yaml_parser_t parser;
+  if (!yaml_parser_initialize(&parser)) {
+    iree_trace_replay_deinitialize(&replay);
+    return iree_make_status(IREE_STATUS_INTERNAL,
+                            "yaml_parser_initialize failed");
+  }
+  yaml_parser_set_input_file(&parser, file);
+
+  iree_status_t status = iree_ok_status();
+  for (bool document_eof = false; !document_eof;) {
+    yaml_document_t document;
+    if (!yaml_parser_load(&parser, &document)) {
+      status = iree_status_from_yaml_parser_error(&parser);
+      break;
+    }
+    yaml_node_t* event_node = yaml_document_get_root_node(&document);
+    if (event_node) {
+      status = iree_trace_replay_event(&replay, &document, event_node);
+    } else {
+      document_eof = true;
+    }
+    yaml_document_delete(&document);
+    if (!iree_status_is_ok(status)) break;
+  }
+
+  yaml_parser_delete(&parser);
+  iree_trace_replay_deinitialize(&replay);
+  return status;
+}
+
+// Runs each of the given traces files sequentially in isolated contexts.
+static iree_status_t iree_run_trace_files(int file_count, char** file_paths,
+                                          iree_vm_instance_t* instance) {
+  for (int i = 0; i < file_count; ++i) {
+    iree_string_view_t file_path = iree_make_cstring_view(file_paths[i]);
+    iree_string_view_t root_path = iree_file_path_dirname(file_path);
+    FILE* file = fopen(file_paths[i], "rb");
+    if (!file) {
+      return iree_make_status(iree_status_code_from_errno(errno),
+                              "failed to open trace file '%.*s'",
+                              (int)file_path.size, file_path.data);
+    }
+    iree_status_t status = iree_run_trace_file(root_path, file, instance);
+    fclose(file);
+    IREE_RETURN_IF_ERROR(status, "replaying trace file '%.*s'",
+                         (int)file_path.size, file_path.data);
+  }
+  return iree_ok_status();
+}
+
+int main(int argc, char** argv) {
```

Comment:
I expected this to be a bit more library-ish, so that we could use it to run benchmarks too. Ok as a starting point but I'd really like benchmark support. Different ways to slice that since you incur some overhead by preparing calls from the yaml...

Reply:
yep - this is just a start - all the iree_trace_replay_* stuff is designed to be hoisted out to another file, but wanted to make sure we had the shape right first